### PR TITLE
Simplify implementation of `run_chain_sync_task`

### PR DESCRIPTION
**Pull request by rafal-ch**
This PR introduces a refactor to some of the functions in `opeartions.rs`, the main goal being the simplification of ~250 lines long `run_chain_sync_task` function. This change will make it easier to add proper metrics to the fast chain sync process (https://github.com/casper-network/casper-node/pull/2640).

The main idea behind the cleanup is to:
* introduce `ChainSyncContext` 
* split the `run_chain_sync_task()` into parts
* wire-up these parts using the context that holds common data needed for processing

### CODE REVIEW

File: node/src/components/chain_synchronizer/operations.rs
Comment (afck) [inline node/src/components/chain_synchronizer/operations.rs]: We should return that, and use it in `fetch_blocks_for_deploy_replay_protection`, otherwise it has an outdated key block info…

File: node/src/components/chain_synchronizer/operations.rs
Comment (afck) [inline node/src/components/chain_synchronizer/operations.rs]: …so we could probably safely use `most_recent_key_block_info` here instead.
(I guess it's not a blocker: We'll just fetch more blocks than necessary this way.)

File: node/src/components/chain_synchronizer/operations.rs
Comment (afck) [inline node/src/components/chain_synchronizer/operations.rs]: Actually, don't we have the trusted _hash_ in the node config? Then we wouldn't need to recompute it here.
Otherwise, maybe we should just add it to `ChainSyncContext`?

File: node/src/components/chain_synchronizer/operations.rs
Comment (afck) [inline node/src/components/chain_synchronizer/operations.rs]: I wonder whether some of these functions should now be `ChainSyncContext` methods instead?
(Doesn't have to be in this PR; just a thought!)

File: node/src/components/chain_synchronizer/operations.rs
Comment (afck) [inline node/src/components/chain_synchronizer/operations.rs:768]: (Then this could be `cts.maybe_key_block_info(most_recent_block.header())`.)

File: node/src/components/chain_synchronizer/operations.rs
Comment (afck) [inline node/src/components/chain_synchronizer/operations.rs:816]: (These functions also all look like `ChainSyncContext` methods now.)

**Review by afck — APPROVED** [review]
Looks good to me! :+1:

File: node/src/components/chain_synchronizer/operations.rs
Comment (rafal-ch) [reply node/src/components/chain_synchronizer/operations.rs]: Resolved in c33e3e886d0af356d770003062fed1a00114292d

File: node/src/components/chain_synchronizer/operations.rs
Comment (rafal-ch) [reply node/src/components/chain_synchronizer/operations.rs]: Resolved in c33e3e886d0af356d770003062fed1a00114292d

File: node/src/components/chain_synchronizer/operations.rs
Comment (rafal-ch) [reply node/src/components/chain_synchronizer/operations.rs]: Yes, we have it, but it's an `Option`.
There is already a code that will read it from the highest block in case it is missing in config and will pass it to `run_chain_sync_task()`: https://github.com/casper-network/casper-node/blob/feat-fast-sync/node/src/components/chain_synchronizer.rs#L74-L86

So yes, I think we should use it in `archival_sync()`. I'll add a commit if all tests pass.

Comment (afck) [reply node/src/components/chain_synchronizer/operations.rs]: Ah, you're right, we need to accept it from either config or storage. Happy to leave it as it is.

Comment (rafal-ch) [reply node/src/components/chain_synchronizer/operations.rs]: Hmm, so we can actually be fetching the same block header twice:
1. in the constructor for `ChainSynchronizer` to read the trusted hash
2. in `run_chain_sync_task()` to read the trusted block header by trusted hash

We could possibly remove the `trusted_hash` parameter from the `start_syncing()` and `run_chain_sync_task()` functions and let the latter read it by itself (from config or from latest block).

Comment (rafal-ch) [reply node/src/components/chain_synchronizer/operations.rs]: The `trusted_hash` is now part of `ChainSyncContext` and is reused: ceea7dba7af6914a226021035a89916d7f299196

File: node/src/components/chain_synchronizer/operations.rs
Comment (rafal-ch) [reply node/src/components/chain_synchronizer/operations.rs]: Yes, that's a good idea. And yes, I'd rather keep it out of the scope of this PR for now.

File: node/src/components/chain_synchronizer/operations.rs
Comment (goral09) [inline node/src/components/chain_synchronizer/operations.rs]: Isn't `trusted_hash == trusted_block_header.hash()`?

Comment (afck) [reply node/src/components/chain_synchronizer/operations.rs]: I hope so! It should not just be the entry from the config, but possibly also our latest block from storage.

Sorry, this was based on my earlier comment, and now I'm not sure about it. If you don't like the duplicated data, we could also make `trusted_hash` a _method_ on `ChainSyncContext`?

Comment (goral09) [reply node/src/components/chain_synchronizer/operations.rs]: It's not about duplication but rather the chance of getting it wrong. If they're equal I think it's just safer to use the hash of the `trusted_block_header`.

Comment (afck) [reply node/src/components/chain_synchronizer/operations.rs]: You're right, I agree. Sorry for the confusion! Let's make it a `trusted_hash` method then.

File: node/src/components/chain_synchronizer/operations.rs
Comment (rafal-ch) [reply node/src/components/chain_synchronizer/operations.rs:768]: See here: https://github.com/casper-network/casper-node/pull/2664#discussion_r807038024

File: node/src/components/chain_synchronizer/operations.rs
Comment (rafal-ch) [reply node/src/components/chain_synchronizer/operations.rs:816]: See here: https://github.com/casper-network/casper-node/pull/2664#discussion_r807038024

File: node/src/components/chain_synchronizer/operations.rs
Comment (rafal-ch) [reply node/src/components/chain_synchronizer/operations.rs]: The `trusted_hash()` method has been added here: https://github.com/casper-network/casper-node/pull/2664/commits/a2de4d36785bd2df35a3dda948986a22029facaf

**Comment by rafal-ch** [comment]
Adding the `do not merge` label, because after the merge of https://github.com/casper-network/casper-node/pull/2650 and fixing the conflicts, the `chain_synchronizer` now uses **all three** "value providers" at the same time:
* chainspec
* chain sync context
* config

I will re-request the review after I clean things up.

**Comment by rafal-ch** [comment]
OK, clean-up finished, re-requesting the review.

File: node/src/components/chain_synchronizer/operations.rs
Comment (afck) [inline node/src/components/chain_synchronizer/operations.rs:168]:
```suggestion
```

File: node/src/components/chain_synchronizer/operations.rs
Comment (afck) [inline node/src/components/chain_synchronizer/operations.rs]: That's done, isn't it?

```suggestion
            // If the new block is a switch block, update the validator weights, etc...
```

**Review by afck — APPROVED** [review]

File: node/src/components/chain_synchronizer/operations.rs
Comment (rafal-ch) [reply node/src/components/chain_synchronizer/operations.rs:168]: Removed https://github.com/casper-network/casper-node/pull/2664/commits/273201177f38a8ee28c5aa2eabe8c7efb03c56e7

File: node/src/components/chain_synchronizer/operations.rs
Comment (rafal-ch) [reply node/src/components/chain_synchronizer/operations.rs]: I see now - if the header contains the "next era validator weights" it means that it's a switch block.
So yes, it's done, indeed.

https://github.com/casper-network/casper-node/pull/2664/commits/1499d9b250d6b4eb5c91f91dd6f30bb4aa6e431e

**Comment by rafal-ch** [comment]
as agreed, merging now with one approval
cc: @Fraser999

**Pull request merged**
